extractor/main.py

```diff
@@ -23,78 +25,164 @@
 )
 from pathlib import Path
 
-# Load the data into a DataFrame
-# climate_action_library_test.csv for testing and changing values
-# climate_action_library_original.csv for original C40 list
-df = load_datafile_into_df(Path("../data/climate_action_library_test.csv"))
-
-# Prepare a list to hold all mapped data
-mapped_data = []
-
-# Define a list of extraction functions and corresponding JSON fields
-extraction_functions = {
-    # "ActionID": extract_ActionID, -> not in the mapping because it is already set in the main script as incremental index
-    "ActionName": extract_ActionName,
-    # "ActionType": extract_ActionType, -> not in the mapping because it is already extracted in the main script as first step
-    "AdaptationCategory": extract_AdaptationCategory,
-    "Hazard": extract_Hazard,
-    # "Sector": extract_Sector,
-    "Subsector": extract_Subsector,
-    "PrimaryPurpose": extract_PrimaryPurpose,
-    "InterventionType": extract_InterventionType,
-    "Description": extract_Description,
-    "BehavioralChangeTargeted": extract_BehavioralChangeTargeted,
-    "CoBenefits": extract_CoBenefits,
-    "EquityAndInclusionConsiderations": extract_EquityAndInclusionConsiderations,
-    "GHGReductionPotential": extract_GHGReductionPotential,
-    "AdaptionEffectiveness": extract_AdaptionEffectiveness,
-    "CostInvestmentNeeded": extract_CostInvestmentNeeded,
-    "TimelineForImplementation": extract_TimelineForImplementation,
-    "Dependencies": extract_Dependencies,
-    "KeyPerformanceIndicators": extract_KeyPerformanceIndicators,
-    "Impacts": extract_Impacts,
-}
-
-# For testing, only process x rows
-df_subset = df  # .head(2)
-
-# Incremental counter for ActionID
-action_id = 1
-
-# Iterate over DataFrame rows
-for index, df_row in df_subset.iterrows():
+# Create a semaphore with a concurrency limit (e.g., 10 tasks at a time)
+semaphore = asyncio.Semaphore(100)
+
+
+async def process_row_with_limit(index, df_row):
+    async with semaphore:  # Limit concurrency
+        return await process_row(index, df_row)
+
+
+async def process_row(index, df_row):
     print(f"Processing row {index}...\n")
     mapped_row = {}
 
     # Assign an incremental value to ActionID
-    mapped_row["ActionID"] = f"{action_id:04d}"
+    # ActionID is row index + 1, zero-padded to 4 digits
+    mapped_row["ActionID"] = f"{index+1:04d}"
 
-    # Extract 'ActionType' first
-    action_type = extract_ActionType(df_row)
+    # Extract synchronous fields
+    action_name = extract_ActionName(index, df_row)
```

Comment:
https://github.com/Open-Earth-Foundation/Climate-Action-Prioritization/blob/57a24175b0d300736c8606a939e36d430d37764b/extractor/utils/extraction_functions.py#L25
will it work if it will be a none or no access to title that you check ? shouldn't it be the other way

Reply:
Inside extract_ActionName I use 

```
action_name = row.get("Title"). 
# Check if the 'Title' column is null or not a string
    if pd.isnull(action_name) or not isinstance(action_name, str):
        action_name = ""
```

So if there is no title it will be None and action_name will be empty string